File: src/spark-examples/sparknormal-examples/SparkStreamingtoHbaseScalaExample010/src/main/scala/com/huawei/bigdata/spark/examples/streaming/SparkOnStreamingToHbase.scala
package com.huawei.bigdata.spark.examples.streaming

import java.io.IOException
import java.util

import org.apache.hadoop.hbase.{HBaseConfiguration, TableName}
import org.apache.hadoop.hbase.util.Bytes
import org.apache.hadoop.hbase.client._
import kafka.serializer.StringDecoder
import org.apache.kafka.common.serialization.StringDeserializer
import org.apache.spark.{SerializableWritable, SparkConf, TaskContext}
import org.apache.spark.streaming.kafka010._
import org.apache.spark.streaming.{Seconds, StreamingContext}

/**
  * run streaming task and select table1 data from hbase, then update to table1
  */
object SparkOnStreamingToHbase {
  def main(args: Array[String]) {
    if (args.length < 3) {
      printUsage
    }

    val Array(checkPointDir, topics, brokers) = args
    val sparkConf = new SparkConf().setAppName("SparkOnStreamingToHbase")
    val ssc = new StreamingContext(sparkConf, Seconds(5))
    val broadcastedConf = ssc.sparkContext.broadcast(new SerializableWritable(ssc.sparkContext.hadoopConfiguration))
    // set CheckPoint dir
    if (!"nocp".equals(checkPointDir)) {
      ssc.checkpoint(checkPointDir)
    }

    val columnFamily = "cf"
    // Create direct kafka stream with brokers and topics
    val topicsSet = topics.split(",").toSet
    // val kafkaParams = Map[String, String]("metadata.broker.list" -> brokers)
    val kafkaParams = scala.collection.Map[String, Object](
      "bootstrap.servers" -> brokers,
      "key.deserializer" -> classOf[StringDeserializer],
      "value.deserializer" -> classOf[StringDeserializer],
      "group.id" -> "testGroup",
      "auto.offset.reset" -> "latest",
      "partition.assignment.strategy" -> "org.apache.kafka.clients.consumer.RangeAssignor",
      "enable.auto.commit" -> (true: java.lang.Boolean)
    )
    // map(_._1) is message key, map(_._2) is message value
    val messages = KafkaUtils.createDirectStream[String, String](
      ssc,
      LocationStrategies.PreferConsistent,
      ConsumerStrategies.Subscribe[String, String](topicsSet, kafkaParams)
    )
    val lines = messages.map(_.value)

    /**
      * do write hbase in executor
      *
      * @param iterator message
      * @param columnFamily columnFamily
      */
    def hBaseWriter(iterator: Iterator[String], columnFamily: String): Unit = {
      val conf = HBaseConfiguration.create(broadcastedConf.value.value)
      var table: Table = null
      var connection: Connection = null

      try {
        connection = ConnectionFactory.createConnection(conf)
        table = connection.getTable(TableName.valueOf("table1"))
        println("connect successful")
        val iteratorArray = iterator.toArray
        val rowList = new util.ArrayList[Get]()
        for (row <- iteratorArray) {
          val get = new Get(row.getBytes)
          rowList.add(get)
        }
        //get data from table1
        val resultDataBuffer = table.get(rowList)
        if (resultDataBuffer.isEmpty) {
          println("fail to get hbase")
        }
        //set data for table1
        val putList = new util.ArrayList[Put]()
        for (i <- 0 until iteratorArray.size) {
          val row = iteratorArray(i)
          val resultData = resultDataBuffer(i)
          if (!resultData.isEmpty) {
            // get value by column Family and colomn qualifier
            val aCid = Bytes.toString(resultData.getValue(columnFamily.getBytes, "cid".getBytes))
            val put = new Put(Bytes.toBytes(row))

            // calculate result value
            val resultValue = row.toInt + aCid.toInt
            put.addColumn(Bytes.toBytes(columnFamily), Bytes.toBytes("cid"), Bytes.toBytes(resultValue.toString))
            putList.add(put)
          }
        }

        if (putList.size() > 0) {
          table.put(putList)
        }
      } catch {
        case e: IOException =>
          e.printStackTrace();
      } finally {
        if (table != null) {
          try {
            table.close()
          } catch {
            case e: IOException =>
              e.printStackTrace();
          }
        }
        if (connection != null) {
          try {
            // Close the HBase connection.
            connection.close()
          } catch {
            case e: IOException =>
              e.printStackTrace()
          }
        }
      }
    }

     // val lines = KafkaUtils.createDirectStream[String, String, StringDecoder, StringDecoder](ssc, kafkaParams, topicSet).map(_._2)
    lines.foreachRDD(rdd => {
      //run for partition
      rdd.foreachPartition(iterator => hBaseWriter(iterator, columnFamily))
    })
    ssc.start()
    ssc.awaitTermination()
  }


  private def printUsage {
    System.out.println("Usage: {checkPointDir} {topic} {brokerList}")
    System.exit(1)
  }
}

File: src/spark-examples/sparknormal-examples/SparkStreamingtoHbaseJavaExample010/src/main/java/com/huawei/bigdata/spark/examples/streaming/SparkOnStreamingToHbase.java
package com.huawei.bigdata.spark.examples.streaming;

import kafka.serializer.StringDecoder;
import scala.Tuple2;

import org.apache.hadoop.conf.Configuration;
import org.apache.hadoop.hbase.*;
import org.apache.hadoop.hbase.client.*;
import org.apache.hadoop.hbase.util.Bytes;
import org.apache.kafka.clients.consumer.ConsumerRecord;
import org.apache.kafka.common.serialization.StringDeserializer;
import org.apache.spark.SparkConf;
import org.apache.spark.api.java.JavaRDD;
import org.apache.spark.api.java.function.Function;
import org.apache.spark.api.java.function.VoidFunction;
import org.apache.spark.streaming.*;
import org.apache.spark.streaming.api.java.JavaDStream;
import org.apache.spark.streaming.api.java.JavaInputDStream;
import org.apache.spark.streaming.api.java.JavaStreamingContext;
import org.apache.spark.streaming.kafka010.*;

import java.io.IOException;
import java.util.*;

/**
 * run streaming task and select table1 data from hbase, then update to table1
 */
public class SparkOnStreamingToHbase {
    public static void main(String[] args) throws Exception {
        if (args.length < 3) {
            printUsage();
        }

        String checkPointDir = args[0];
        String topics = args[1];
        final String brokers = args[2];

        Duration batchDuration = Durations.seconds(5);
        SparkConf sparkConf = new SparkConf().setAppName("SparkOnStreamingToHbase");
        JavaStreamingContext jssc = new JavaStreamingContext(sparkConf, batchDuration);

        // set CheckPoint dir
        if (!"nocp".equals(checkPointDir)) {
            jssc.checkpoint(checkPointDir);
        }

        final String columnFamily = "cf";
        Map<String, Object> kafkaParams = new HashMap<String, Object>();
        kafkaParams.put("bootstrap.servers", brokers);
        kafkaParams.put("key.deserializer", StringDeserializer.class);
        kafkaParams.put("value.deserializer", StringDeserializer.class);
        kafkaParams.put("group.id", "testGroup");
        kafkaParams.put("auto.offset.reset", "latest");
        kafkaParams.put("partition.assignment.strategy", "org.apache.kafka.clients.consumer.RangeAssignor");
        kafkaParams.put("enable.auto.commit", true);

        String[] topicArr = topics.split(",");
        Set<String> topicSet = new HashSet<String>(Arrays.asList(topicArr));

        // Create direct kafka stream with brokers and topics
        JavaInputDStream<ConsumerRecord<String, String>> stream =
                KafkaUtils.createDirectStream(
                        jssc,
                        LocationStrategies.PreferConsistent(),
                        ConsumerStrategies.<String, String>Subscribe(topicSet, kafkaParams));
        JavaDStream<String> lines =
                stream.map(
                        new Function<ConsumerRecord<String, String>, String>() {
                            public String call(ConsumerRecord<String, String> v1) throws Exception {
                                return v1.value();
                            }
                        });

        lines.foreachRDD(
                new VoidFunction<JavaRDD<String>>() {
                    public void call(JavaRDD<String> rdd) throws Exception {
                        rdd.foreachPartition(
                                new VoidFunction<Iterator<String>>() {
                                    public void call(Iterator<String> iterator) throws Exception {
                                        hBaseWriter(iterator, columnFamily);
                                    }
                                });
                    }
                });

        jssc.start();
        jssc.awaitTermination();
    }

    /**
     * do write hbase in executor
     *
     * @param iterator     message
     * @param columnFamily columnFamily
     */
    private static void hBaseWriter(Iterator<String> iterator, String columnFamily) throws IOException {
        Configuration conf = HBaseConfiguration.create();
        Connection connection = null;
        Table table = null;
        try {
            connection = ConnectionFactory.createConnection(conf);
            table = connection.getTable(TableName.valueOf("table1"));
            System.out.println("connect sucessfully");
            List<Get> rowList = new ArrayList<Get>();
            while (iterator.hasNext()) {
                Get get = new Get(iterator.next().getBytes());
                rowList.add(get);
            }
            // get data from table1
            Result[] resultDataBuffer = table.get(rowList);

            // set data for table1
            List<Put> putList = new ArrayList<Put>();
            for (int i = 0; i < resultDataBuffer.length; i++) {
                String row = new String(rowList.get(i).getRow());
                Result resultData = resultDataBuffer[i];
                if (!resultData.isEmpty()) {
                    // get value by column Family and colomn qualifier
                    String aCid = Bytes.toString(resultData.getValue(columnFamily.getBytes(), "cid".getBytes()));
                    Put put = new Put(Bytes.toBytes(row));

                    // calculate result value
                    int resultValue = Integer.valueOf(row) + Integer.valueOf(aCid);
                    put.addColumn(
                            Bytes.toBytes(columnFamily),
                            Bytes.toBytes("cid"),
                            Bytes.toBytes(String.valueOf(resultValue)));
                    putList.add(put);
                }
            }

            if (putList.size() > 0) {
                table.put(putList);
            }
        } catch (IOException e) {
            e.printStackTrace();
        } finally {
            if (table != null) {
                try {
                    table.close();
                } catch (IOException e) {
                    e.printStackTrace();
                }
            }
            if (connection != null) {
                try {
                    // Close the HBase connection.
                    connection.close();
                } catch (IOException e) {
                    e.printStackTrace();
                }
            }
        }
    }

    private static void printUsage() {
        System.out.println("Usage: {checkPointDir} {topic} {brokerList}");
        System.exit(1);
    }
}
